pkg/trace/force_sampler.go

```diff
@@ -71,8 +66,27 @@ func (s *otelForceSampler) ShouldSample(p sdktrace.SamplingParameters) sdktrace.
 	}
 }
 
-func forceSample(sampleRate uint) sdktrace.Sampler {
+// isSampled is used to determine if the given trace ID should be sampled based on the
+// current sample rate.
+//
+// Please keep the sampling logic deterministic because this method is used to correlate sampling between different
+// services at Outreach. For example, if a producer of the trace headers generates traceID1 and a remote service
+// wants to link the traceID1 to its own trace (!= traceID1), then it should only do so if traceID1 was sampled.
+// Failure to do so causes honeycomb to show a trace link to a 'dead trace' - and it is not possible to filter those
+// out, rendering the the cross-trace linking experience useless.
+func (s *otelForceSampler) isSampled(traceID string) bool {
+	sampler, err := sample.NewDeterministicSampler(s.sampleRate)
```

Comment:
While it is true that `sample.NewDeterministicSampler` is stateless, we should probably still not allocate this on each sample call.  I know the earlier code used to do the same but can we take the chance to fix this by adding the sampler as state to otelForceSampler struct (as part of the newSampler call).

Also, I'd move the panic to the `newSampler` call rather than have panic in the `shouldSample` code path.

Reply:
+1 here

Reply:
sure, I have moved the init logic with the panic into the newSampler in the last iteration